quine-core/src/main/scala/com/thatdot/quine/graph/NodeActor.scala

```diff
@@ -139,6 +134,73 @@ private[graph] class NodeActor(
     case MergedHere(other) => !mergedIntoHere.contains(other)
   }
 
+  /**
+    * Drop duplicate events, and wrap the Events being kept in an `EventTime`.
+    *
+    * An event is considered a duplicate (and is dropped) if:
+    *  - the same event appears later in the sequence of events, or
+    *  - the event would have no effect when applied to this node.
+    *
+    * This implementation prioritizes avoiding heap allocations.
+    * Since it does a linear search of the later events,
+    * this could be less efficient for very large sequences of events.
+    *
+    * @param events An ordered sequence of events to be de-duplicated.
+    * @param atTimeOverride If defined, the `EventTime` that should be combined with the single event;
+    *                       If empty, the an `EventTime` is obtained using `nextEventTime()`.
+    * @return a sequence of de-duplicated events wrapped in a `WithTime`.
+    *         This result is mutable, so it should not be allowed to escape this context.
+    */
+  def deduplicateEvents(events: Seq[NodeChangeEvent],
+                        atTimeOverride: Option[EventTime]
+                       ): Seq[WithTime] = {
+    require(atTimeOverride.isEmpty || events.length == 1, "Can only override EventTime for a single Event")
+
+    val r: ArrayBuffer[WithTime] = new ArrayBuffer[WithTime](events.length)
+
+    @tailrec
+    def accumulate(events: Seq[NodeChangeEvent] = events): Seq[WithTime] =
+      events match {
+        case event +: laterEvents =>
```

Comment:
Just to circle back to my previous comment, and for the sake of easy reading of the review: here is the outer loop over `events` that lifts the `exists` predicates from "linear" to "quadratic" at the level of `deduplicateEvents`.